### PR TITLE
PL-127 text selection misalignment for reference view

### PR DESCRIPTION
**Pull request by Giamburrasca**
DONE:
- Text rendered
- No more misalignment
- Scroll&zoom blocked
- Blocked the close button on the top
- Refactor the methods

### CODE REVIEW

File: src/components/ReferenceViewComponent.js [hunk 1]
```diff
@@ -122,44 +137,75 @@ class ReferenceViewComponent {
...
+    
   };
+
+  /**
+ * Function to render the page of the reference
+ * @param {pdfDoc} pdfDoc PDF document
+ * @param {Object} component object that holds DOM elements that are within component
+ * @param {HTMLElement} pageNumber number of the page of the reference selected
+ */
+#renderPageReference = (pdfDoc, component, pageNumber) => {
```
Comment (NinaGnjidic) [inline src/components/ReferenceViewComponent.js:163]: Like I said on discord, content of this method is repeated so many times in our code, so it should be extracted in some kind of service and code repetition should be eliminated... If you manage to do this that would be great.

**Review by NinaGnjidic — COMMENTED** [review]
Tested locally, when I open first pdf everything works as expected, I open a new one, open a reference and this happens:
image

File: src/components/ReferenceViewComponent.js (hunk 1 above)
Comment (Giamburrasca) [reply src/components/ReferenceViewComponent.js]: Done

File: src/components/PdfPageComponent.js [hunk 2]
```diff
@@ -60,9 +60,9 @@ class PdfPageComponent {
 
     let page = await this.pdfDoc.getPage(this.pageNum);
 
-    let viewport = this.#setupViewport(page, zoomScale);
+    let viewport = getViewport(page, this.components.canvas, zoomScale);
```
Comment (tonesija) [inline src/components/PdfPageComponent.js:63]: What about setupViewport method? It should be deleted.

**Review by tonesija — COMMENTED** [review]
But what about reusing PdfPageComponent for ReferenceView?

File: src/components/ReferenceViewComponent.js [hunk 3]
```diff
@@ -27,14 +34,16 @@ class ReferenceViewComponent {
     pdfContainer: document.querySelector("#pdf-container"),
     contentDiv: document.querySelector("#content-reference-div"),
     sidePageReferenceBtn: document.querySelector("#side-page-reference-btn"),
-    sidePageReferenceContainer: document.createElement("canvas"),
+    sidePageReferenceContainer: document.createElement("div"),
     main: document.querySelector("#main"),
     pdfDoc: null,
     pdfPageNumber: null,
     viewport: null,
```
Comment (tonesija) [inline src/components/ReferenceViewComponent.js:41]: I know this is not your code but this should not be in components.

File: src/services/TextRenderService.js [hunk 4]
```diff
@@ -30,14 +30,20 @@ export function hideLinks() {
  *
  * @param {int} pageNum
  */
-function createTextLayerDOMIfNotExist(pageNum) {
+function createTextLayerDOMIfNotExist(pageNum, reference = false) {
```
Comment (tonesija) [inline src/services/TextRenderService.js:33]: What is reference? Missing documentation.

Comment (tonesija) [reply src/services/TextRenderService.js:33]: This is not good, these should be 2 separeted functions.

File: src/services/TextRenderService.js [hunk 5]
```diff
@@ -132,3 +138,66 @@ function renderLinkLayer(page, textLayer, viewport) {
...
+  const ctx = canvas.getContext("2d");
+  const renderCtx = {
+    canvasContext: ctx,
+    viewport: viewport,
+  };
+  return renderCtx;
+}
+
+/**
+ * Creates the page viewport and sets canvas size.
```
Comment (tonesija) [inline src/services/TextRenderService.js:159]: It is a little weird to have a "get" function mutate objects.

File: src/services/TextRenderService.js [hunk 6]
```diff
@@ -132,3 +138,66 @@ function renderLinkLayer(page, textLayer, viewport) {
...
+  return viewport;
+}
+
+  /**
+ * Function to render the page inside the reference side view
+ * @param {pdfDoc} pdfDoc PDF document
+ * @param {Object} component object that holds DOM elements that are within component
+ * @param {HTMLElement} pageNumber number of the page of the reference selected
+ */
+export function renderPageReference(pdfDoc, component, pageNumber) {
```
Comment (tonesija) [inline src/services/TextRenderService.js:182]: Try to use async/await

File: src/components/PdfPageComponent.js (hunk 2 above)
Comment (Giamburrasca) [reply src/components/PdfPageComponent.js:63]: Done

File: src/services/TextRenderService.js (hunk 4 above)
Comment (Giamburrasca) [reply src/services/TextRenderService.js:33]: Done, separated in two functions!

File: src/services/TextRenderService.js (hunk 5 above)
Comment (Giamburrasca) [reply src/services/TextRenderService.js:159]: That's the way to have small functions working for both the render.

**Comment by gabrielekarra** [comment]
pushed the design, btw we have to include this in sidepagecomponent @Giamburrasca as knowledge graph and summary

**Comment by Giamburrasca** [comment]
> pushed the design, btw we have to include this in sidepagecomponent @Giamburrasca as knowledge graph and summary

We will do everything but test before pushing, now there are more errors than before!
The text is misalignment again :[

IMAGE 2023-01-05 11:40:43

**Comment by gabrielekarra** [comment]
> > pushed the design, btw we have to include this in sidepagecomponent @Giamburrasca as knowledge graph and summary
> 
> We will do everything but test before pushing, now there are more errors than before! The text is misalignment again :[
> 
> IMAGE 2023-01-05 11:40:43

i tested, for me still works

**Comment by gabrielekarra** [comment]
+1, others changes can be done in refactor task

File: src/components/ReferenceViewComponent.js (hunk 3 above)
Comment (Giamburrasca) [reply src/components/ReferenceViewComponent.js]: Done

**Comment by paolocorsa** [comment]
For me is ok.
+1.

**Review by matteovisotto — APPROVED** [review]
All fine for me, +1

**Pull request merged**